### PR TITLE
[river] Ensure that the WS transport always calls the close callback

### PR DESCRIPTION
**Pull request by lhchavez**
## Why

The UDS transport has a separate error and close callbacks, but apparently for some reason the WS one only calls one.

[We see](https://app.datadoghq.com/logs?query=%40replid%3Aa127f237-08ec-4184-a6b7-763e008cbff8&agg_m=count&agg_m_source=base&agg_t=count&cols=core_host%2Ccore_service&event=AgAAAY_zGqbGIETeNQAAAAAAAAAYAAAAAEFZX3pHdmVIQUFBb3pXaE9iM1BYS0FBQQAAACQAAAAAMDE4ZmYzMWItM2UzZi00N2I3LTk2YTYtYmI3MzI5OTE5YmU2&fromUser=true&messageDisplay=inline&refresh_mode=sliding&storage=hot&stream_sort=desc&viz=stream&from_ts=1717098390703&to_ts=1717703190703&live=true) that there are browsers that, upon an unclean exit, calls the `onerror` callback, but Node doesn't. This is why our tests always pass!

## What changed

This change now always calls both the error and close callbacks. Mostly because client-side WS errors out _a lot_ (any non-1000 close code is considered an error). This means that now a network-driven disconnect will allow us to reconnect before the deadline elapses.

## Versioning

- [ ] Breaking protocol change
- [ ] Breaking ts/js API change

### CODE REVIEW

**Review by masad-frost — APPROVED** [review]
Feels like it might be a better approach to call connect in the error callback https://github.com/replit/river/blob/cf0ecd1a60d6dd0df23ed3cb73c2ddbc20d2c41d/transport/transport.ts#L637? Relying on the semantics of calling close after error feels fragile

**Comment by masad-frost** [comment]
> Feels like it might be a better approach to call connect in the error callback
> 
> https://github.com/replit/river/blob/cf0ecd1a60d6dd0df23ed3cb73c2ddbc20d2c41d/transport/transport.ts#L637
> 
> ? Relying on the semantics of calling close after error feels fragile

Actually nevermind, if uds and websocket behave the same way then we can make it an expectation

File: transport/impls/ws/ws.test.ts [hunk 1]
```diff
@@ -170,4 +170,40 @@ describe('network edge cases', async () => {
...
+    const msg1 = createDummyTransportMessage();
+    const msg2 = createDummyTransportMessage();
+
+    const msg1Id = clientTransport.send(serverTransport.clientId, msg1);
+    await expect(
+      waitForMessage(serverTransport, (recv) => recv.id === msg1Id),
+    ).resolves.toStrictEqual(msg1.payload);
+
+    // unclean server disconnect. Note that the Node implementation sends the reason on the
+    // `onclose`, but (some?) browsers call the `onerror` handler before it since it was an unclean
```
[the remark this answers is not in the document]
Comment (lhchavez) [reply transport/impls/ws/ws.test.ts:197]: well... got bad news for you, haha.

[the remark this answers is not in the document]
Comment (lhchavez) [reply transport/impls/ws/ws.test.ts:197]: or rather, it might be the case that what node considers to be "an error" differs from browsers' POV. i suspect that node will call `onerror` if the conn did drop, but not if the conn was cleanly closed (with an error).
